server/remote_cache/content_addressable_storage_server/content_addressable_storage_server.go

```diff
@@ -265,12 +284,57 @@ func (s *ContentAddressableStorageServer) BatchReadBlobs(ctx context.Context, re
 			blobRsp.Status = &statuspb.Status{Code: int32(codes.Internal)}
 		} else {
 			blobRsp.Status = &statuspb.Status{Code: int32(codes.OK)}
+			if s.env.GetConfigurator().GetCacheZstdTranscodingEnabled() && clientAcceptsCompressor(req.AcceptableCompressors, repb.Compressor_ZSTD) {
+				blobRsp = zstdCompressResponse(blobRsp, req.AcceptableCompressors)
+			}
 		}
+
 		rsp.Responses = append(rsp.Responses, blobRsp)
 	}
 	return rsp, nil
 }
 
+func (s *ContentAddressableStorageServer) supportsCompressor(compressor repb.Compressor_Value) bool {
+	return compressor == repb.Compressor_IDENTITY ||
+		compressor == repb.Compressor_ZSTD && s.env.GetConfigurator().GetCacheZstdTranscodingEnabled()
+}
+
+func clientAcceptsCompressor(acceptableCompressors []repb.Compressor_Value, compressor repb.Compressor_Value) bool {
+	// Per protocol, IDENTITY is always accepted.
+	if compressor == repb.Compressor_IDENTITY {
+		return true
+	}
+	for _, c := range acceptableCompressors {
+		if c == compressor {
+			return true
+		}
+	}
+	return false
+}
+
+func zstdCompressResponse(blobRsp *repb.BatchReadBlobsResponse_Response, acceptableCompressors []repb.Compressor_Value) *repb.BatchReadBlobsResponse_Response {
+	data, err := zstd.Compress(nil, blobRsp.Data)
+	if err != nil {
+		log.Errorf("Failed to compress blob: %s", err)
+		return &repb.BatchReadBlobsResponse_Response{
+			Digest: blobRsp.Digest,
+			Status: gstatus.Convert(err).Proto(),
```

Comment:
is this error coming from zstd? might want to wrap in an InternalErrorf before converting to proto

Reply:
Good catch, fixed